internal/resolver/resolver.go

```diff
@@ -578,13 +601,43 @@ func (r *resolver) loadAsFileOrDirectory(path string) (string, bool) {
 	return "", false
 }
 
+func resolvePathWithoutStar(from, path string) (string, error) {
+	replaced := strings.Replace(path, "/*", "", -1)
+	return fp.Join(from, replaced), nil
+}
+
 func (r *resolver) loadNodeModules(path string, dirInfo *dirInfo) (string, bool) {
 	for {
 		// Handle TypeScript base URLs for TypeScript code
-		if dirInfo.tsConfigJson != nil && dirInfo.tsConfigJson.absPathBaseUrl != nil {
-			basePath := r.fs.Join(*dirInfo.tsConfigJson.absPathBaseUrl, path)
-			if absolute, ok := r.loadAsFileOrDirectory(basePath); ok {
-				return absolute, true
+		if dirInfo.tsConfigJson != nil {
+
+			if dirInfo.tsConfigJson.absPathBaseUrl != nil {
+				if dirInfo.tsConfigJson.paths != nil {
+					for key, originalPaths := range dirInfo.tsConfigJson.paths {
+						for _, originalPath := range originalPaths {
+							if matched, err := regexp.MatchString("^"+key, path); matched && err == nil {
```

Comment:
Using a regex to match this isn't correct. The pattern is an asterisk, which causes a regex to repeat the previous character zero or more times. That would mean a pattern `foo*` would also match a directory named `foooo` and `fo`. I took some time to look at what the TypeScript compiler itself does. The relevant code is in `tryLoadModuleUsingPaths()` [here](https://github.com/microsoft/TypeScript/blob/c194abb38c54c1607831b4ec678c0bdd15f7fe49/src/compiler/moduleNameResolver.ts#L1361-L1387).

Other relevant code is `tryParsePattern()` [here](https://github.com/microsoft/TypeScript/blob/c194abb38c54c1607831b4ec678c0bdd15f7fe49/src/compiler/utilities.ts#L6122-L6130):

```ts
    export function tryParsePattern(pattern: string): Pattern | undefined {
        // This should be verified outside of here and a proper error thrown.
        Debug.assert(hasZeroOrOneAsteriskCharacter(pattern));
        const indexOfStar = pattern.indexOf("*");
        return indexOfStar === -1 ? undefined : {
            prefix: pattern.substr(0, indexOfStar),
            suffix: pattern.substr(indexOfStar + 1)
        };
    }
```

and `isPatternMatch()` [here](https://github.com/microsoft/TypeScript/blob/c194abb38c54c1607831b4ec678c0bdd15f7fe49/src/compiler/core.ts#L2015-L2019):

```ts
    function isPatternMatch({ prefix, suffix }: Pattern, candidate: string) {
        return candidate.length >= prefix.length + suffix.length &&
            startsWith(candidate, prefix) &&
            endsWith(candidate, suffix);
    }
```

Reply:
thank you! i've added some commits to handle this more like ts compiler, but i haven't written the test case where `*` is at the start or middle of the path.